### ISSUE TITLE
Apple siliconb DrRacket.app not opening

### ISSUE THREAD

**Issue by bestlem** [racket/racket#3666]
**What version of Racket are you using?**
e.g., 8.0.0.1
 DrRacket, version 8.0.0.1--2021-01-16(b09e10d/a) [cs].
**What program did you run?**
Just double click on DrRackey or from terminal `open DrRacket.app`

**What should have happened?**
DRRacket opens and you can run ythings

**If you got an error message, please include it here.**
If I do 
`open --stderr ~/tmp/err --stdout ~/tmp/out 'DrRacket.app'`
~/t.err contains
```
standard-module-name-resolver: collection not found
  for module path: (lib "drracket/drracket.rkt")
  collection: "drracket"
  in collection directories:
   /Users/mark/Library/Racket/8.0.0.1/collects
```

**Please include any other relevant details**
macOS 11.1 Apple Silicon similar non starting for other apps in the release.

However if I run from the terminal the executable inside DrRacket.app e.g.
`'/Applications/Racket v8.0.0.1/DrRacket.app/Contents/MacOS/DrRacket'`

It all works

**Comment by mflatt** [comment]
Thanks for the report!

I think this must be a Gatekeeper interaction that prevents "DrRacket.app" seeing adjacent files. When I try that installer (looks like it's from the Northwestern snapshot site), it works on my machine (11.2 on Apple Silicon). But we've seen these kinds of problems in the past where installations behave differently on different machines. The problem had seemed to go away with the introduction of notarization, though.

What does
```
xattr -l Racket\ v8.0.0.1/
```
and
```
xattr -r Racket\ v8.0.0.1/DrRacket.app 
```
show for you?

For me, it's
```
com.apple.quarantine: 0181;601566e9;Firefox;C84DC6D8-2A14-49BE-851F-6E1BD4A04CF3
```
and
```
com.apple.quarantine: 01c1;601566e9;Firefox;C84DC6D8-2A14-49BE-851F-6E1BD4A04CF3
```
I think the 0181 vs. 01c1 difference is because I've run "DrRacket.app".

I also wonder is whether a snapshot from the Utah site works. There shouldn't be any difference, but it may be helpful to know whether this is any difference on your machine.

**Comment by bestlem** [comment]
/Applications🔒 xattr -l Racket\ v8.0.0.1/
com.apple.quarantine: 0081;6003106e;Firefox;2707B1D7-5FCB-40C6-8469-9CF282D21121

and I think you meant -l in the second case as well
Applications🔒 xattr -l Racket\ v8.0.0.1/DrRacket.app
com.apple.quarantine: 00c1;6003106e;Firefox;2707B1D7-5FCB-40C6-8469-9CF282D21121
 and I don't think the difference is because youi ran DrRacket.app as I have not run the games
xattr -l Racket\ v8.0.0.1/PLT\ Games.app
com.apple.quarantine: 00c1;6003106e;Firefox;2707B1D7-5FCB-40C6-8469-9CF282D21121

**Comment by bestlem** [comment]
On the latest downloads 8.0.0.3 is NW and 8.0.0.4 is Utah
/Applications🔒 xattr -l Racket\ v8.0.0.3
com.apple.quarantine: 0081;60158a7e;Firefox;4A199D3E-DAD4-4D73-A4D7-7CAA4A546E1C
xattr -l Racket\ v8.0.0.3/DrRacket.app
com.apple.quarantine: 0081;60158a7e;Firefox;4A199D3E-DAD4-4D73-A4D7-7CAA4A546E1C
and open gets the same message as before 
and 
xattr -l Racket\ v8.0.0.3/DrRacket.app
com.apple.quarantine: 00c1;60158a7e;Firefox;4A199D3E-DAD4-4D73-A4D7-7CAA4A546E1C

SO 91 to c1 is opening.
xattr -l Racket\ v8.0.0.4
com.apple.quarantine: 0081;60158a54;Firefox;1F10C32A-ABDA-4F7C-978E-235F12DF441B
xattr -l Racket\ v8.0.0.4/DrRacket.app
com.apple.quarantine: 0081;60158a54;Firefox;1F10C32A-ABDA-4F7C-978E-235F12DF441B
and open gets the same error
xattr -l Racket\ v8.0.0.4/DrRacket.app
com.apple.quarantine: 00c1;60158a54;Firefox;1F10C32A-ABDA-4F7C-978E-235F12DF441B

**Comment by bestlem** [comment]
Still not working in the release
standard-module-name-resolver: collection not found
  for module path: (lib "drracket/drracket.rkt")
  collection: "drracket"
  in collection directories:
   /Users/mark/Library/Racket/8.0/collects

I would note that I do not have the directory /Users/mark/Library/Racket

So perhaps it is an install error

**Comment by mflatt** [comment]
Sorry — I lost track of this one. I'm pretty sure the error is because Mac OS thinks the application needs to be sandboxed, and the sandbox prevents DrRacket from finding adjacent libraries (because the sandbox confines access to the ".app" bundle). But I don't know why it gets sandboxed on your machine.

If you use something like `xattr -r -d Racket\ v8.0/DrRacket.app`, does that fix the problem?

**Comment by bestlem** [comment]
Yes xattr -dr com.apple.quarantine does fix the issue. 

I think Racket either needs an installer - or The files are made part of the app bundle

**Comment by jbclements** [comment]
I don't think that everyone that installs Racket on Apple silicon has the same unfortunate experience as you. Specifically, it appears to me that the signatures applied to the bundle as part of Apple's code-signing process mean that for most users, the application is not sandboxed as it was for you. I admit that I have no idea why your installation was sandboxed.

**Comment by saclark** [comment]
I'm actually having a the same issue, though perhaps for a slightly differnt reason, running on a 2020 Intel (_not M1_) MacBook Pro.

In my case, I already had `minimal-racket` (version 7.9) installed via homebrew and a working verion of DrRacket that I had installed via `raco pkg install --auto drracket`. I then upgraded `minimal-racket` via homebrew and ran `raco pkg install --auto drracket` to install DrRacket. However, attemptinmg to run DrRacket fails due to the same error as previously discussed:
```
standard-module-name-resolver: collection not found
  for module path: (lib "drracket/drracket.rkt")
  collection: "drracket"
  in collection directories:
   /Users/sclark/Library/Racket/8.0/collects
   /usr/local/Cellar/minimal-racket/8.0/share/racket/collects/
```
In my case, however, `ls /usr/local/Cellar/minimal-racket/8.0/share/racket/collects` shows that no such `drracket` directory exists (and `/Users/sclark/Library/Racket/8.0/collects` does not actually exist on my system). I'm not sure why that lib would be missing.

As a sidenote, in case it is relevant, I also have troubles upgrading `minimal-racket` with homebrew bc it upgrades my racket version but fails to update my `config.rktd` in `/usr/local/etc/racket`. What happens is that the values specified in that file still refer to my old version (e.g. `(catalogs . ("https://download.racket-lang.org/releases/7.9/catalog/" #f)` instead of `(catalogs . ("https://download.racket-lang.org/releases/8.0/catalog/" #f)`)). As a result, DrRacket also fails to install correctly until I correct this (because it is looking in the wrong catalog). However, a correct `config.rktd.default` file that references my new verion _is_ created when I upgrade `minimal-racket`, so I simply copy the contents of `config.rktd.default` to `config.rktd` to correct it. Note, that I had done this to correct my `config.rktd` before installing DrRacket.

**Comment by bestlem** [comment]
And 8.1 has the same issue

**Comment by koratkar** [comment]
Has this issue been resolved?

**Comment by bestlem** [comment]
No I have the same issue on 8.2. However from other issues some machines have it some don't

**Comment by shhyou** [comment]
Which `.dmg` file did you download? Is it a snapshot build?

**Comment by bestlem** [comment]
@shhyou  For 8.0 8.1 and 8.2 the release build - for original case various 8.0.0.x builds

**Comment by shhyou** [comment]
I see. In Systems -> Security & Privacy, are apps from identified developers allowed?
![security-preference](https://user-images.githubusercontent.com/1910898/131505137-c68f1e47-24c2-41aa-b3b8-f095c8f24bf6.png)

**Comment by bestlem** [comment]
Yes

**Comment by bestlem** [comment]
I note the error is similar to that if you haven't run  ```raco pkg migrate 8.1```

I first got when I tried to run one of my programs for the command line 
 bench.rkt:4:1: collection not found
  for module path: benchmark
  collection: "benchmark"
  in collection directories:
   /Users/mark/Library/Racket/8.2/collects
   /Applications/Racket v8.2/collects/
   ... [168 additional linked and package directories]
  location...:
   bench.rkt:4:1

**Comment by ogoletti** [comment]
Hi, 
I also have a similar issue on MacOS 12.2.1 (M1). 

I installed `minimal-racket` through `brew install minimal-racket` then I also installed `drracket`with `raco pkg install --auto drracket`. 

The app will not launch, and while in `/opt/homebrew/Cellar/minimal-racket/8.4/bin` doing `./drracket` or `./DrRacket.app/Contents/MacOS/DrRacket` both give me the error : 
```
standard-module-name-resolver: collection not found
  for module path: (lib "drracket/drracket.rkt")
  collection: "drracket"
  in collection directories:
   /Users/ogoletti/Library/Racket/8.4/collects
   /opt/homebrew/Cellar/minimal-racket/8.4/share/racket/collects/
```

**Comment by soegaard** [comment]
Hi,

A quick fix is to use the official installer instead:

    https://download.racket-lang.org/
    
That said, it would good to know what causes the issue.

**Comment by ogoletti** [comment]
I guess that will install a new racket installation somewhere on my computer ? 
I can launch drracket via `racket drracket.rkt` in `/opt/homebrew/Cellar/minimal-racket/8.4/share/racket/pkgs/drracket/drracket` too.
Tks

**Comment by bestlem** [comment]
@soeguard - no the official installer does not work - see my posts here. Also 8.4 still does not work

**Comment by ogoletti** [comment]
@soegaard  @bestlem I just installed the M1 dmg `racket-8.4-aarch64-macosx-cs.dmg`  from the website and it works as-is. So I can confirm that workaround works too for me.

**Comment by bestlem** [comment]
Racket 8.5 fails in the same way

stderr is

standard-module-name-resolver: collection not found
  for module path: (lib "drracket/drracket.rkt")
  collection: "drracket"
  in collection directories:
   /Users/mark/Library/Racket/8.5/collects

**Comment by soegaard** [comment]
Two observations:

1.  Since the problem persists - maybe there is a problem in a local configuration file?

2. If you used the official installer, why is folder not

    /Applications/Racket v8.5/collects
    
?
[Do you have that folder too?]

**Comment by sorawee** [comment]
Well, that line makes sense. If you run `(require this-module-do-not-exist)`, you will get something like `/Users/mark/Library/Racket/8.5/collects` too. 

I presume that `/Applications/Racket v8.5/collects/` is on the second line, but that didn't get copied over.

**Comment by bestlem** [comment]
> I presume that `/Applications/Racket v8.5/collects/` is on the second line, but that didn't get copied over.

No - it does not appear.

**Comment by bestlem** [comment]
> Two observations:
> 
>     1. Since the problem persists - maybe there is a problem in a local configuration file?
> 
>     2. If you used the official installer, why is folder not
>        /Applications/Racket v8.5/collects
> 
> 
> ? [Do you have that folder too?]

That directory does not appear in stderr but it does exist on my machine and yes I would expect Drracet to have found this..

I do have gatekeeper set to App Store and other developers - but if that fails I should get an Apple dialog.
I do get before running DrRacket from open or the desktop an Apple dialog saying that this is downloaded from the internet and do I want to run it.

Also note if I run DrRacket from the command line directly it did run OK - this time I did that before running from open or the desktop. That sort of behaviour in other cases has been due to trying to use the current directory which is / if via open or desktop but in a shell it is usial Unix (today it was `Applications/Racket v8.5` (Oh please don't provide a space )

**Comment by bestlem** [comment]
Currently there are more details here the discourse is just a reference to it.,
Going forward this thread is discoverable bvy Goole, Bing at al and not hidden so will be seen by more.

**Comment by jbclements** [comment]
I think I don't understand your comment, are you suggesting it's a mistake to link discourse posts to github issues?

**Comment by ixnij** [comment]
I found that `xattr -r -d Racket\ v8.0/DrRacket.app` didn't affect in my case, and I got some error instead:
```console
/Applications🔒
❯ xattr -r -d Racket\ v8.0/DrRacket.app
Not enough arguments for option -d. Expected at least 2 but got 1

usage: xattr [-l] [-r] [-s] [-v] [-x] file [file ...]
       xattr -p [-l] [-r] [-s] [-v] [-x] attr_name file [file ...]
       xattr -w [-r] [-s] [-x] attr_name attr_value file [file ...]
       xattr -d [-r] [-s] attr_name file [file ...]
       xattr -c [-r] [-s] file [file ...]

The first form lists the names of all xattrs on the given file(s).
The second form (-p) prints the value of the xattr attr_name.
The third form (-w) sets the value of the xattr attr_name to the string attr_value.
The fourth form (-d) deletes the xattr attr_name.
The fifth form (-c) deletes (clears) all xattrs.

options:
  -h: print this help
  -l: print long format (attr_name: attr_value and hex output has offsets and
      ascii representation)
  -r: act recursively
  -s: act on the symbolic link itself rather than what the link points to
  -v: also print filename (automatic with -r and with multiple files)
  -x: attr_value is represented as a hex string for input and output
```

**Comment by jbclements** [comment]
The help text you quoted suggests to me it may be the case that `xattr` now requires you to explicitly name the attribute you wish to remove. What do you get from 

```
xattr -rd com.apple.quarantine Racket\ v8.5.900
```

?

**Comment by jbclements** [comment]
See also 

https://github.com/racket/racket/wiki/Configure-Command-Line-for-Racket

(which makes the same suggestion about the use of xattr, I believe)

**Comment by ixnij** [comment]
> The help text you quoted suggests to me it may be the case that `xattr` now requires you to explicitly name the attribute you wish to remove. What do you get from
> 
> ```
> xattr -rd com.apple.quarantine Racket\ v8.5.900
> ```
> 
> ?

```console
❯ xattr -rd com.apple.quarantine Racket\ v8.5.900
❯ echo $?
0
```
I got nothing, but it seems that it's successful: I can now open DrRacket by clicking the icon in launchpad or via the open command.

**Comment by jbclements** [comment]
Sorry, yes, poorly worded on my part.

This confirms that your problem was also a quarantine problem. I still don't know what causes this, though.

**Comment by ixnij** [comment]
I'm sorry that I forget to tell you that the dmg file was originally downloaded by my mobile phone via LTE, as my broadband is too slow. I transferred the dmg file to my computer via an application named WeChat, it might be the reason of my failure. So when I run the command `xattr -l Racket\ v8.5.900/DrRacket.app` for the first time, I got `com.apple.quarantine: 00c2;62e14d02;WeChat;`. As you can see, it must be related with WeChat.

**Comment by jbclements** [comment]
Oh! That could certainly explain it.

**Comment by e0Xs9x8i** [comment]
I'm having the same issue except I can't run racket.app from terminal. When I do

> xattr -rd com.apple.quarantine Racket\ v8.5 

it doesn't say anything nor does it let me open

**Comment by sorawee** [comment]
@e0Xs9x8i Perhaps try Racket 8.6?

**Comment by Inc0n** [comment]
I am not sure why, but `drracket/drracket.rkt` is found in `/pkgs` and not in `/collects` which is where drracket.app is looking for. Could this be something that's related to the cause of the issue?

```
┌─14:59 /opt/homebrew/Cellar/minimal-racket/8.7/share/racket/collects
└─√ $ fd drracket.rkt -p /opt/homebrew/Cellar/minimal-racket/8.7/share/racket/ 

/opt/homebrew/Cellar/minimal-racket/8.7/share/racket/pkgs/drracket/drracket/compiled/drracket_rkt.dep
/opt/homebrew/Cellar/minimal-racket/8.7/share/racket/pkgs/drracket/drracket/compiled/drracket_rkt.zo
/opt/homebrew/Cellar/minimal-racket/8.7/share/racket/pkgs/drracket/drracket/drracket.rkt

┌─15:01 /opt/homebrew/Cellar/minimal-racket/8.7/share/racket/collects
└─√ $ drracket 
standard-module-name-resolver: collection not found
  for module path: (lib "drracket/drracket.rkt")
  collection: "drracket"
  in collection directories:
   /Users/xception/Library/Racket/8.7/collects
   /opt/homebrew/Cellar/minimal-racket/8.7/share/racket/collects/
```

I am using m1 and installed `minimal-racket` using `brew`. `drracket` is installed using `raco`.

**Comment by Inc0n** [comment]
> Out of curiosity: is there a reason that you used `brew install minimal-racket` rather than `brew install racket` ?

I have been using emacs for the time being, therefore, I did not think I need the full drracket experience yet.
Which is why I installed `minimal-racket`. However, I have been looking for something greater than emacs, something I desire so especially after watching the talk "Stop Writing Dead Programs" by Jack Rusher on strangeloop. 

It is a sort of literate programming similar to org mode, except more interactive, which is where org mode lacks. So it may save me some time just to ask, do you reckon `drracket` would be the right place to look into this?

**Secondly**, as it turns out `(require drracket)` works in racket repl and starts drracket just fine. But running `drracket` directly on the cmd line does not.

**Comment by Inc0n** [comment]
> For answers about DrRacket & its affordances vis a vis Jack Rusher's talk, I strongly encourage you to ask on the discourse group, https://racket.discourse.group/ ; I bet you'll get some good replies there.

Right, got it. 

> WRT starting drracket, what do you get from
> 
> `racket -l drracket`

That works too. Same as `(require drracket)` in racket repl.

**Comment by jbclements** [comment]
How about double-clicking on the DrRacket application?

**Comment by Inc0n** [comment]
> How about double-clicking on the DrRacket application?

It wouldn't start. And again what I shown in https://github.com/racket/racket/issues/3666#issuecomment-1371043982 was the result of running `drracket`, which is a short script that invokes `.app`, therefore, please have a look at it for more information.

**Comment by jbclements** [comment]
Right, the question was whether the problem was in the script or in the actual app bundle. Based on your response, it does not appear that there's any problem with the script.

I'm relatively confident that the problem here is connected to the way that homebrew installs its formulae. Indeed, it seems that the "Cask" mechanism is designed specifically for GUI apps, so it may simply be the case that the easiest way to fix this is to uninstall the minimal-racket formula and install the racket cask, using `brew install racket`.

**Comment by bestlem** [comment]
It is not just Homebrew. Using Racket supplied .dmg gives the same problem. So an issue with Racket not necessarily Homebrew.

And downloading by iPhone is not needed. Just download on the Mac you are installing to.
